Modification for main modules compatibility

szmazurek opened the pull request:
Fixes #816

## Proposed Changes
- modification for proper scheduler and optimizer handling
- modification for proper onnx handling
- GAN logger instance
- adding model to the models registry

## Checklist

- [X] [`CONTRIBUTING`](https://github.com/mlcommons/GaNDLF/blob/master/CONTRIBUTING.md) guide has been followed.
- [X] PR is based on the [current GaNDLF master ](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch-in-github-desktop?platform=windows).
- [X] Non-breaking change (does **not** break existing functionality): provide **as many** details as possible for _any_ breaking change.
- [X] Function/class source code documentation added/updated (ensure `typing` is used to provide type hints, including and not limited to using `Optional` if a variable has a pre-defined value).
- [X] Code has been [blacked](https://github.com/psf/black#usage) for style consistency and linting.
- [ ] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/version.py).
- [ ] If adding a git submodule, add to list of exceptions for black styling in [pyproject.toml](https://github.com/mlcommons/GaNDLF/blob/master/pyproject.toml) file.
- [ ] [Usage documentation](https://github.com/mlcommons/GaNDLF/blob/master/docs) has been updated, if appropriate.
- [ ] Tests added or modified to [cover the changes](https://app.codecov.io/gh/mlcommons/GaNDLF); if coverage is reduced, please give explanation.
- [ ] If customized dependency installation is required (i.e., a separate `pip install` step is needed for PR to be functional), please ensure it is reflected in all the files that control the CI, namely: [python-test.yml](https://github.com/mlcommons/GaNDLF/blob/master/.github/workflows/python-test.yml), and all docker files [[1](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-CPU),[2](https://github.com/mlcommons/GaNDLF/blob/devcontainer_build_fix/Dockerfile-CUDA11.6),[3](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-ROCm)].

szmazurek committed:
Modification for main modules compatibility

- modification for proper scheduler and optimizer handling
- modification for proper onnx handling
- GAN logger instance
- adding model to the models registry

sarthakpati commented on GANDLF/models/__init__.py:
```diff
@@ -35,6 +45,7 @@
 from .brain_age import brainage
 from .unetr import unetr
 from .transunet import transunet
+from .dcgan import DCGAN
```
Not sure if this is needed for **this** PR...?

szmazurek replied:
nah, you are right. Will move it into the dcgan branch

sarthakpati replied:
Thanks!

sarthakpati commented on GANDLF/models/__init__.py:
```diff
@@ -105,6 +116,7 @@
     "msdnet": MSDNet,
     "brain_age": brainage,
     "sdnet": SDNet,
+    "dcgan": DCGAN,
```
Not sure if this is needed for **this** PR...?

szmazurek replied:
as above

sarthakpati reviewed (commented):
Most of it is fine. Just 1 question to address.

szmazurek committed:
Update __init__.py

Moving imports into dcgan branch

szmazurek committed:
Update __init__.py

Moving model imports to dcgan branch

szmazurek commented:
> Most of it is fine. Just 1 question to address.

Waiting for tests to pass, them will merge.
EDIT tests will not pass as there is an error due to generic function modification I made (I think I mentioned that it will fail, I expected that). Can I merge and then adress that?

sarthakpati commented on GANDLF/utils/tensor.py:
```diff
@@ -101,7 +108,9 @@ def reverse_one_hot(predmask_tensor: torch.Tensor, class_list: Union[List[int],
-                if case in _class:  # check if any of the special cases are present
+                if (
+                    case in _class
+                ):  # check if any of the special cases are present
```
Run `black` through this again.
```suggestion
                # check if any of the special cases are present
                if (
                    case in _class
                ):
```

sarthakpati commented on GANDLF/data/ImagesFromDataFrame.py:
```diff
@@ -93,7 +92,12 @@ def ImagesFromDataFrame(
+        subject_id,
+        channel_str,
+        loader_type,
+        extension,
```
```suggestion
        resized_image: str,
        output_dir: str,
        subject_id: str,
        channel_str: str,
        loader_type: str,
        extension: str,
```

sarthakpati commented on GANDLF/logger.py:
```diff
@@ -71,3 +76,83 @@ def write(
+            numeric_input (float or torch.Tensor): The input value
+        Returns:
+            str: The parsed input value
+        """
```
Update comment style.

szmazurek replied:
Done

sarthakpati commented on GANDLF/logger.py:
```diff
@@ -71,3 +76,83 @@ def write(
+            epoch_number (int): The epoch number
+            disc_loss (float or torch.Tensor): The discriminator loss
+            gen_loss (float or torch.Tensor): The generator loss
+            epoch_metrics (dict): The epoch metrics
```
Update comment style.

szmazurek replied:
Done

sarthakpati commented on GANDLF/utils/tensor.py:
```diff
@@ -139,15 +152,20 @@ def send_model_to_device(
+        optimizer_1 (torch.optim): The optimizer for training. In case of GANs, this
+    is the optimizer for the generator.
+        optimizer_2 (torch.optim): The second optimizer, used only in case of GANs.
+    This is the optimizer for the discriminator.
```
```suggestion
        optimizer_1 (torch.optim): The optimizer for training. In case of GANs, this is the optimizer for the generator.
        optimizer_2 (torch.optim): The second optimizer, used only in case of GANs. This is the optimizer for the discriminator.
```

sarthakpati commented on GANDLF/utils/tensor.py:
```diff
@@ -204,9 +223,11 @@ def send_model_to_device(
 
-        if not (optimizer is None):
+        if not (optimizer_1 is None):
             # ensuring optimizer is in correct device - https://github.com/pytorch/pytorch/issues/8741
-            optimizer.load_state_dict(optimizer.state_dict())
+            optimizer_1.load_state_dict(optimizer_1.state_dict())
+        if not (optimizer_2 is None):
+            optimizer_2.load_state_dict(optimizer_2.state_dict())
```
Wondering if it would be better to have a list or dict of optimizers instead of just 2...?

szmazurek replied:
It is one of possible solutions, this was the first thing I come up with. Should I add that to this pr or we add it in the future?

sarthakpati replied:
Do you mind adding it now so that are _kind of_ future proofed? 😆

szmazurek replied:
Added that, will still need to modify it in one place for GANs but that after the merge

sarthakpati reviewed (changes requested):
What's the version of `black` you are using, BTW? Is it different from what's in [`setup.py`](https://github.com/mlcommons/GaNDLF/blob/master/setup.py#L85)?

Also, minor changes.

szmazurek commented:
> What's the version of `black` you are using, BTW? Is it different from what's in [`setup.py`](https://github.com/mlcommons/GaNDLF/blob/master/setup.py#L85)?
> 
> Also, minor changes.

I have the same, but probably I have different line length settings - I am in the 79 length clan, I believe default is 120 line length? Should I change that back?

sarthakpati commented:
> > What's the version of `black` you are using, BTW? Is it different from what's in [`setup.py`](https://github.com/mlcommons/GaNDLF/blob/master/setup.py#L85)?
> > Also, minor changes.
> 
> I have the same, but probably I have different line length settings - I am in the 79 length clan, I believe default is 120 line length? Should I change that back?

How are you suggesting black to make that change? I am curious because I don't think GaNDLF is setting that in the linting configuration...

szmazurek commented:
> > > What's the version of `black` you are using, BTW? Is it different from what's in [`setup.py`](https://github.com/mlcommons/GaNDLF/blob/master/setup.py#L85)?
> > > Also, minor changes.
> > 
> > 
> > I have the same, but probably I have different line length settings - I am in the 79 length clan, I believe default is 120 line length? Should I change that back?
> 
> How are you suggesting black to make that change? I am curious because I don't think GaNDLF is setting that in the linting configuration...

I am setting that in the vscode black plugin settings, I have it configured that vscode automatically formats the document upon saving

sarthakpati commented:
> > > > What's the version of `black` you are using, BTW? Is it different from what's in [`setup.py`](https://github.com/mlcommons/GaNDLF/blob/master/setup.py#L85)?
> > > > Also, minor changes.
> > > 
> > > 
> > > I have the same, but probably I have different line length settings - I am in the 79 length clan, I believe default is 120 line length? Should I change that back?
> > 
> > 
> > How are you suggesting black to make that change? I am curious because I don't think GaNDLF is setting that in the linting configuration...
> 
> I am setting that in the vscode black plugin settings, I have it configured that vscode automatically formats the document upon saving

Okay - do you mind setting it to the default (i.e., 120)? It's just that if this is not a parameter that we are controlling on the black configuration for GaNDLF, these changes will come up for each contributor... 🤕

szmazurek commented:
> > > > > What's the version of `black` you are using, BTW? Is it different from what's in [`setup.py`](https://github.com/mlcommons/GaNDLF/blob/master/setup.py#L85)?
> > > > > Also, minor changes.
> > > > 
> > > > 
> > > > I have the same, but probably I have different line length settings - I am in the 79 length clan, I believe default is 120 line length? Should I change that back?
> > > 
> > > 
> > > How are you suggesting black to make that change? I am curious because I don't think GaNDLF is setting that in the linting configuration...
> > 
> > 
> > I am setting that in the vscode black plugin settings, I have it configured that vscode automatically formats the document upon saving
> 
> Okay - do you mind setting it to the default (i.e., 120)? It's just that if this is not a parameter that we are controlling on the black configuration for GaNDLF, these changes will come up for each contributor... 🤕

I do not - will do that! And actually, default turns out to be 88 chars/line, so this is what I am going to set

szmazurek committed:
Update GANDLF/utils/tensor.py

Co-authored-by: Sarthak Pati <patis@iu.edu>

szmazurek committed:
Update GANDLF/data/ImagesFromDataFrame.py

add typing

Co-authored-by: Sarthak Pati <patis@iu.edu>

szmazurek committed:
Updated typing and docstrings

szmazurek committed:
Modified moving objects to device

- function send_model_to_device.py now takes a list of optimizers and
  places them at correct device
- calls to the functions were modified, tests working

sarthakpati commented on GANDLF/utils/tensor.py:
```diff
@@ -328,7 +333,10 @@ def get_class_imbalance_weights_segmentation(
     # iterate through full penalty data
     for _, (subject) in enumerate(
-        tqdm(penalty_loader, desc="Looping over training data for penalty calculation")
+        tqdm(
```
Is the lines for `black` 120 or 79 for this file?

szmazurek replied:
It was with the default, so 88, I mentioned that in the last comment in the previous conversation. Should I go with 120 instead? BTW - I re-blacked once again all files in this pr, pushing now

szmazurek replied:
Ha! Strange thing, I manually moved it back to the version with this code in single line black did not move it back to the previous format. Dunno why I had to do it manually

sarthakpati reviewed (commented):
Looks good. Just confirm what the number of characters are for black for these files (it seems like it is less than what it was before).

szmazurek committed:
Black all PR files once again

sarthakpati commented on GANDLF/logger.py:
```diff
@@ -71,3 +71,93 @@ def write(
+
+            for metric in epoch_metrics:
+                row += self._parse_input_value(epoch_metrics[metric]) + ","
+            row = row[:-1]  # ??
```
What is this comment about? 😆

szmazurek replied:
I was wondering what this line does :P But I managed to understand why it is like that, I think I will remove it :P

szmazurek replied:
removed

sarthakpati commented on GANDLF/utils/tensor.py:
```diff
@@ -483,8 +488,16 @@ def print_model_summary(
+            "\tParameters size:",
+            stats.to_megabytes(stats.total_param_bytes),
+            "MB",
+        )
```
Seems like the line length is <88 for this block.

szmazurek replied:
It was like 1 character (last comma in the brackets) that was preventing that from being set to single-line version. Removed it, also I have installed black from main branch, will use this one now.

sarthakpati replied:
Oh okay. The change doesn't seem to have worked for all files, though. For example:

![image](https://github.com/szmazurek/GaNDLF/assets/11719673/ca7d3eb7-9bd7-4171-a39a-ecd089923f7c)

Can you confirm?

szmazurek replied:
Look -  when I ran the black from CMD, there was no difference from the file I have commited here. However, some lines  got changed when I passed the -C flag, which stands for "don't go to newline if there is a trailing comma" (you can see it with black --help). Regarding the line you mentioned, take a look at my editor - I have a ruler set for 88 chars, and this one is slightly over that:
![image](https://github.com/szmazurek/GaNDLF/assets/48680634/55850f77-198a-4a06-9495-ab84cf05d906)

szmazurek replied:
Pushed the files ran with black with the commands I mentioned, give them a look and let's see if it did the trick. If so, maybe the -C flag should go into main branch for gandlf in extension config?

sarthakpati replied:
> -C flag should go into main branch for gandlf in extension config

Do you know how to set this up on the config? If so, can you please put in the PR?

szmazurek replied:
Short answer - no idea, but will give it a shot

szmazurek replied:
Edit - there are cases online where people set that in toml. I did it and installed correctly, however vscode does not give a damn about that for some reason. Explicitly setting this in the vscode ext settings works, but from cmd I still need the -C flag. No idea how to get over that, I have separate branch ready if you still want it to be commited

sarthakpati replied:
Yes, let's put that in as a PR. Thanks!

szmazurek replied:
Ok done. Closing then, if nothing elese then I am merging

sarthakpati reviewed (changes requested):
I don't think the `black` line-length setting was set properly. Perhaps this should be done using a project-level setting.

sarthakpati commented:
> I don't think the `black` line-length setting was set properly. Perhaps this should be done using a project-level setting.

[This PR](https://github.com/mlcommons/GaNDLF/pull/820) should set this up correctly.

szmazurek committed:
Black re-formatting

szmazurek committed:
Black formatting

- formatted all files from this pr using black -l 88 -C command
- now I think the lines are set correctly
